### PR TITLE
PHP: add php server interop test

### PR DESCRIPTION
**Pull request by HannahShiSFB**
@stanley-cheung

### CODE REVIEW

File: src/php/tests/interop/interop_server.php [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        list($initial_metadata, $trailing_metadata) =
+            $this->maybeEchoMetadata($context);
+        return [
+            new \Grpc\Testing\EmptyMessage(),
+            $initial_metadata,
+            \Grpc\Status::ok($trailing_metadata)
+        ];
+    }
+
+    public function UnaryCall(
```
Comment (stanley-cheung) [inline src/php/tests/interop/interop_server.php:73]: Holy cow this is awesome!

File: src/php/lib/Grpc/RpcServer.php [hunk 2]
```diff
@@ -32,137 +32,120 @@
...
+            try {
+                $this->processCall(
+                    $method_desc,
+                    $server_reader,
+                    $server_writer,
+                    $context
+                );
+            } catch (\Exception $e) {
+                $server_writer->finish(
+                    Status::status(GRPC_STATUS_INTERNAL, $e->getMessage())
```
Comment (stanley-cheung) [inline src/php/lib/Grpc/RpcServer.php:95]: Oh you already have the `INTERNAL` status here. Oh so the constructor can deal with all the statuses. Then I guess there's no need for those convenience methods for OK and Unimplemented.

File: src/php/lib/Grpc/RpcServer.php [hunk 3]
```diff
@@ -32,137 +32,120 @@
...
-        $namespace = $rf->getParentClass()->getNamespaceName();
-        $prefix = "";
-        if ($namespace) {
-            $parts = explode("\\", $namespace);
-            foreach ($parts as $part) {
-                $prefix .= lcfirst($part) . ".";
-            }
+    public function handle($service)
+    {
+        $MethodDescs = $service->getMethodDescriptors();
```
[the remark this answers is not in the document]
Comment (HannahShiSFB) [reply src/php/lib/Grpc/RpcServer.php:52]: comment addressed

File: src/php/lib/Grpc/RpcServer.php (hunk 2 above)
Comment (HannahShiSFB) [reply src/php/lib/Grpc/RpcServer.php:95]: OK and Unimplemented are used in generated stub

File: src/php/bin/generate_proto_php.sh [hunk 4]
```diff
@@ -40,7 +40,7 @@ mv $output_file ./src/proto/grpc/testing/test.proto
 # interop test protos
 $PROTOC --proto_path=. \
        --php_out=src/php/tests/interop \
-       --grpc_out=src/php/tests/interop \
+       --grpc_out=generate_server:src/php/tests/interop \
```
[the remark this answers is not in the document]
Comment (HannahShiSFB) [reply src/php/bin/generate_proto_php.sh:43]: as we talked offline, we'll keep this flag as false by default, and opt-in when generate a server

File: src/php/ext/grpc/completion_queue.c [hunk 5]
```diff
@@ -27,4 +27,4 @@ void grpc_php_init_completion_queue(TSRMLS_D) {
 void grpc_php_shutdown_completion_queue(TSRMLS_D) {
   grpc_completion_queue_shutdown(completion_queue);
   grpc_completion_queue_destroy(completion_queue);
-}
+}
```
[the remark this answers is not in the document]
Comment (HannahShiSFB) [reply src/php/ext/grpc/completion_queue.c:30]: comment addressed

File: src/php/lib/Grpc/RpcServer.php [hunk 6]
```diff
@@ -32,137 +32,120 @@
...
+                    $context
+                );
+            } catch (\Exception $e) {
+                $server_writer->finish(
+                    Status::status(GRPC_STATUS_INTERNAL, $e->getMessage())
+                );
             }
+        } catch (\Exception $e) {
+            fwrite(STDERR, "ERROR: " . $e->getMessage() . PHP_EOL);
+            exit(1);
```
[the remark this answers is not in the document]
Comment (HannahShiSFB) [reply src/php/lib/Grpc/RpcServer.php:100]: revisit

File: src/php/lib/Grpc/Status.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    public static function ok(array $metadata = null): array
+    {
+        return Status::status(STATUS_OK, 'OK', $metadata);
+    }
+    public static function unimplemented(): array
+    {
+        return Status::status(STATUS_UNIMPLEMENTED, 'UNIMPLEMENTED');
+    }
+}
```
[the remark this answers is not in the document]
Comment (HannahShiSFB) [reply src/php/lib/Grpc/Status.php:55]: only these 2 are used for now, will add more when necessary

File: src/php/lib/Grpc/RpcServer.php [hunk 8]
```diff
@@ -32,137 +32,120 @@
...
-        if ($namespace) {
-            $parts = explode("\\", $namespace);
-            foreach ($parts as $part) {
-                $prefix .= lcfirst($part) . ".";
-            }
+    public function handle($service)
+    {
+        $MethodDescs = $service->getMethodDescriptors();
+        $exist_methods = array_intersect_key($this->paths_map, $MethodDescs);
+        if (!empty($exist_methods)) {
```
[the remark this answers is not in the document]
Comment (HannahShiSFB) [reply src/php/lib/Grpc/RpcServer.php:54]: comment addressed. override added.

**Pull request closed without merging**